### PR TITLE
Bug fix: PubSub tests "overrun" their timers (XPlattCpp)

### PR DESCRIPTION
**Pull request by JTonoJimenez**

### CODE REVIEW

File: targets/XPlatCppSdk/source/test/TestApp/TestContext.cpp [hunk 1]
```diff
@@ -8,6 +8,29 @@ namespace PlayFabUnit
 {
     void TestContext::EndTest(TestFinishState state, std::string resultMsg)
     {
+        if (finishState != TestFinishState::PENDING)
+        {
+            resultMsg += testResultMsg;
```
Comment (JTonoJimenez) [inline targets/XPlatCppSdk/source/test/TestApp/TestContext.cpp]: Saves the previous Message saved, just to know if we pass or fail before.

File: targets/XPlatCppSdk/source/test/TestApp/TestContext.cpp [hunk 2]
```diff
@@ -8,6 +8,29 @@ namespace PlayFabUnit
 {
     void TestContext::EndTest(TestFinishState state, std::string resultMsg)
     {
+        if (finishState != TestFinishState::PENDING)
```
Comment (JTonoJimenez) [inline targets/XPlatCppSdk/source/test/TestApp/TestContext.cpp]: If the status is not Pending that means that we already finish the test at least once, and we are running it again  for a different reason.

File: targets/XPlatCppSdk/source/test/TestApp/TestContext.cpp [hunk 3]
```diff
@@ -8,6 +8,29 @@ namespace PlayFabUnit
...
+        {
+            resultMsg += testResultMsg;
+            resultMsg += "You can't mark a test as finished twice.\n";
+            if (finishState == TestFinishState::PASSED)
```
Comment (JTonoJimenez) [inline targets/XPlatCppSdk/source/test/TestApp/TestContext.cpp]: This are the cases where we want to handle of sending a new message or adding times.

File: targets/XPlatCppSdk/source/test/TestApp/TestContext.cpp [hunk 4]
```diff
@@ -8,6 +8,29 @@ namespace PlayFabUnit
...
+            else
+            {
+                if (state != TestFinishState::FAILED)
+                {
+                    resultMsg += "You can't mark a test as finished twice.\n";
+                }
+                testResultMsg += resultMsg;
+                return;
+            }
+        }
```
Comment (pgilmorepf) [inline targets/XPlatCppSdk/source/test/TestApp/TestContext.cpp]: It's really hard to follow this logic.

I would prefer:

if (condition && condition)
// logic
else if (condition && condition)
// logic
else if (condition && condition)
// logic
else if (condition
// logic
else
// logic

Basically, your logic above may be right, but if exactly one state is wrong, then we have to reinvent this whole code. The switch statement I describe above may be more code, but it's a lot easier to debug and add a condition pair that may not be covered by your logic.

**Comment by Vicodex** [comment]
retest

**Comment by Vicodex** [comment]
retest

**Comment by Vicodex** [comment]
Jorge, this PR fails on Xbox (test application times out). Without these changes, Xbox succeeds. It means something in this PR is not working as intended.

**Comment by pgilmorepf** [comment]
retest

**Comment by pgilmorepf** [comment]
retest
